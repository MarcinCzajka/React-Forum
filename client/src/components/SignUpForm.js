import React from 'react';
import { Message, Button, Form, Modal, Menu } from 'semantic-ui-react';
import basePath from '../api/basePath';
import UserContext from '../contexts/UserContext';
import { UserConsumer } from '../contexts/UserContext';
import Captcha from './Captcha';

class SignUpForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            'open': false,
            'userName': '',
            'email': '',
            'password': '',
            'avatar': '',
            'error': ''
        }
    }

    static contextType = UserContext;

    login = () => {
        if(this.state.password !== this.state.passwordRepeat) {
            this.setState({error: 'Passwords are not identical.'})
            return;
        }

        basePath({
            method: 'post',
            url: '/api/users',
            data: {
                name: this.state.userName,
                email: this.state.email,
                password: this.state.password
            }
        }).then(res => {
            if (res.status === 200) {

                this.setState({
                    'open': false
                })

                alert('Account created.')
            }
        }).catch(error => {
                console.log(error);
                this.setState({ error: error.response.data });
            });
    }

    error = () => {
        if (this.state.error) {
            return (
                <Message error
                header = 'Error'
                content = {this.state.error} />
            )
        }
    }

    render() {
        return (
            <UserConsumer>
                {context => (
                    <div>
                        {!context.loggedIn ? (
                            <Modal size='tiny' trigger={<Menu.Item onClick={this.open} name='Sign up!' />}
                                open={this.state.open}
                                onClose={this.close}
                            >
                                <Modal.Header>Sign up</Modal.Header>
                                <Modal.Content>
                                    <Form size='small' onSubmit={this.login} error >
                                        <div>{this.error()}</div>
                                        <br></br>
                                        <Form.Field required>
                                            <label>Username</label>
                                            <input placeholder='Name' value={this.state.userName} onChange={ (e) => this.setState({ userName: e.target.value })} />
                                        </Form.Field>
                                        <Form.Field required>
                                            <label>Email</label>
                                            <input placeholder='Email' value={this.state.email} onChange={ (e) => this.setState({ email: e.target.value })} />
                                        </Form.Field>
                                        <Form.Field required>
                                            <label>Password</label>
                                            <input 
                                                type='password'
                                                placeholder='Password'
                                                spellcheck = "false"
                                                autocapitalize = "off"
                                                autocorrect = "off"
                                                value={this.state.password} 
                                                onChange={(e) => this.setState({ password: e.target.value })} 
                                            />
                                            <input 
                                                type='password'
                                                placeholder='Repeat'
                                                spellcheck = "false"
                                                autocapitalize = "off"
                                                autocorrect = "off"
                                                value={this.state.passwordRepeat} 
                                                onChange={(e) => this.setState({ passwordRepeat: e.target.value })} 
                                            />
                                        </Form.Field>

                                        <Captcha />

                                        <Button type='submit' fluid size='large'>Sign up</Button>
                                    </Form>
                                </Modal.Content>
                            </Modal>
                        ) : ''}
                    </div>
                )}
            </UserConsumer>
        )
    }

    close = () => {
        this.setState({open: false});
    }
    open = () => {
        this.setState({open: true});
    }

}

export default SignUpForm;