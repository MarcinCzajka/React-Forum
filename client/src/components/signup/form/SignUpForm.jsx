import React from 'react';
import { Button, Form, Modal } from 'semantic-ui-react';
import Captcha from './Captcha';
import ErrorMessage from '../../message/ErrorMessage';
import { LocaleConsumer } from '../../../contexts/LocaleContext';

class SignUpForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            'userName': '',
            'email': '',
            'password': '',
            'passwordRepeat': '',
            'captchaToken': ''
        }
    }

    verifyCaptcha = (captchaToken) => {
        this.setState({captchaToken: captchaToken})
    }

    render() {
        const { 
            open, 
            onClose,
            submit,
            errorMessage
        } = this.props;

        return (
            <LocaleConsumer>
                {locale => (
                    <Modal 
                        size='tiny'
                        open={open}
                        onClose={onClose}
                    >
                        <Modal.Header>{locale.signUp.modalHeader}</Modal.Header>
                        <Modal.Content>
                            <Form 
                                size='small'
                                onSubmit={() => submit(this.state)}
                                error
                            >

                                <ErrorMessage message={errorMessage} />

                                <Form.Field required>
                                    <label>{locale.signUp.label.username}</label>
                                    <input 
                                        value={this.state.userName}
                                        autoComplete='off'
                                        placeholder={locale.signUp.label.usernamePlaceholder}
                                        onChange={(e) => this.setState({ userName: e.target.value })} 
                                    />
                                </Form.Field>

                                <Form.Field required>
                                    <label>{locale.signUp.label.email}</label>
                                    <input 
                                        value={this.state.email}
                                        autoComplete="new-password"
                                        placeholder={locale.signUp.label.emailPlaceholder}
                                        onChange={(e) => this.setState({ email: e.target.value })}
                                    />
                                </Form.Field>

                                <Form.Field required>
                                    <label>{locale.signUp.label.password}</label>
                                    <input 
                                        type='password'
                                        placeholder={locale.signUp.label.passwordPlaceholder}
                                        spellCheck = "false"
                                        autoCapitalize = "off"
                                        autoComplete="new-password"
                                        autoCorrect = "off"
                                        value={this.state.password} 
                                        onChange={(e) => this.setState({ password: e.target.value })} 
                                    />
                                    <input 
                                        type='password'
                                        placeholder={locale.signUp.label.passrepeatPlaceholder}
                                        spellCheck = "false"
                                        autoCapitalize = "off"
                                        autoComplete="new-password"
                                        autoCorrect = "off"
                                        value={this.state.passwordRepeat} 
                                        onChange={(e) => this.setState({ passwordRepeat: e.target.value })} 
                                    />
                                </Form.Field>

                                <Form.Field>
                                    <Captcha verifyCallback={this.verifyCaptcha} />
                                </Form.Field>

                                <Button type='submit' fluid size='large'>{locale.signUp.modalHeader}</Button>

                            </Form>

                        </Modal.Content>
                    </Modal>
                )}
            </LocaleConsumer>
        )
    }

}

export default SignUpForm;