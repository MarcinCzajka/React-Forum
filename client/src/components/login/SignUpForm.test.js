import React from 'react';
import { create } from "react-test-renderer";
import SignUpForm from './SignUpForm';
import { UserProvider } from '../../contexts/UserContext';

jest.mock('semantic-ui-react/dist/commonjs/addons/Portal/Portal', () => ({ children }) => children);

const loggedIn = bool => {
    return {
        loggedIn: bool,
        addShowSignupToContext: function(){}
    }
}

describe('SignUpForm component', () => {
    test('match snapshot w/o user logged in', () => {
        const component = create(<UserProvider value={loggedIn(false)}><SignUpForm /></UserProvider>);
        expect(component).toMatchSnapshot();
    });

    test('match snapshot /w user logged in', () => {
        const component = create(<UserProvider value={loggedIn(true)}><SignUpForm /></UserProvider>);
        expect(component).toMatchSnapshot();
    });

});