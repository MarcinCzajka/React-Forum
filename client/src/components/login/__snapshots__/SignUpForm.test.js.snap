// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm component match snapshot /w user logged in 1`] = `
<div>
  
</div>
`;

exports[`SignUpForm component match snapshot w/o user logged in 1`] = `
<div>
  <div
    className="ui page modals dimmer transition visible active"
  >
    <div
      className="ui tiny modal transition visible active"
      style={Object {}}
    >
      <div
        className="header"
      >
        Sign up
      </div>
      <div
        className="content"
        style={
          Object {
            "paddingTop": 0,
          }
        }
      >
        <form
          className="ui small error form"
          onSubmit={[Function]}
        >
          <div />
          <br />
          <div
            className="required field"
          >
            <label>
              Username
            </label>
            <input
              autoComplete="off"
              onChange={[Function]}
              placeholder="Name"
              value=""
            />
          </div>
          <div
            className="required field"
          >
            <label>
              Email
            </label>
            <input
              autoComplete="new-password"
              onChange={[Function]}
              placeholder="Email"
              value=""
            />
          </div>
          <div
            className="required field"
          >
            <label>
              Password
            </label>
            <input
              autoCapitalize="off"
              autoComplete="new-password"
              autoCorrect="off"
              onChange={[Function]}
              placeholder="Password"
              spellCheck="false"
              type="password"
              value=""
            />
            <input
              autoCapitalize="off"
              autoComplete="new-password"
              autoCorrect="off"
              onChange={[Function]}
              placeholder="Repeat"
              spellCheck="false"
              type="password"
            />
          </div>
          <div
            className="field"
          >
            <div
              data-onloadcallbackname="onloadCallback"
              data-verifycallbackname="verifyCallback"
              id="g-recaptcha"
            />
          </div>
          <button
            className="ui large fluid button"
            onClick={[Function]}
            type="submit"
          >
            Sign up
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
